src/java.base/share/classes/jdk/internal/util/DecimalDigits.java

```diff
@@ -136,4 +131,306 @@ public static int stringSize(long x) {
+     */
+    public static int getChars(long i, int index, char[] buf) {
+        // Used by trusted callers.  Assumes all necessary bounds checks have been done by the caller.
+        long q;
+        int charPos = index;
+
+        boolean negative = (i < 0);
+        if (!negative) {
+            i = -i;
+        }
+
+        // Get 2 digits/iteration using longs until quotient fits into an int
+        while (i <= Integer.MIN_VALUE) {
+            q = i / 100;
+            charPos -= 2;
+            putPair(buf, charPos, (int)((q * 100) - i));
+            i = q;
+        }
+
+        // Get 2 digits/iteration using ints
+        int q2;
+        int i2 = (int)i;
+        while (i2 <= -100) {
+            q2 = i2 / 100;
+            charPos -= 2;
+            putPair(buf, charPos, (q2 * 100) - i2);
+            i2 = q2;
+        }
+
+        // We know there are at most two digits left at this point.
+        if (i2 < -9) {
+            charPos -= 2;
+            putPair(buf, charPos, -i2);
+        } else {
+            buf[--charPos] = (char) ('0' - i2);
+        }
+
+        if (negative) {
+            buf[--charPos] = '-';
+        }
+        return charPos;
+    }
+
+    /**
+     * Insert the 2-chars integer into the buf as 2 decimal digit ASCII chars,
+     * only least significant 16 bits of {@code v} are used.
+     * @param buf byte buffer to copy into
+     * @param charPos insert point
+     * @param v to convert
+     */
+    public static void putPair(char[] buf, int charPos, int v) {
+        int packed = DIGITS[v];
+        buf[charPos    ] = (char) (packed & 0xFF);
+        buf[charPos + 1] = (char) (packed >> 8);
+    }
+
+    /**
+     * Insert the 2-bytes integer into the buf as 2 decimal digit ASCII bytes,
+     * only least significant 16 bits of {@code v} are used.
+     * @param buf byte buffer to copy into
+     * @param charPos insert point
+     * @param v to convert
+     */
+    public static void putPairLatin1(byte[] buf, int charPos, int v) {
+        int packed = DIGITS[v];
+        putCharLatin1(buf, charPos, packed & 0xFF);
+        putCharLatin1(buf, charPos + 1, packed >> 8);
+    }
+
+    /**
+     * Insert the 2-chars integer into the buf as 2 decimal digit UTF16 bytes,
+     * only least significant 16 bits of {@code v} are used.
+     * @param buf byte buffer to copy into
+     * @param charPos insert point
+     * @param v to convert
+     */
+    public static void putPairUTF16(byte[] buf, int charPos, int v) {
+        int packed = DIGITS[v];
+        putCharUTF16(buf, charPos, packed & 0xFF);
+        putCharUTF16(buf, charPos + 1, packed >> 8);
+    }
+
+    private static void putCharLatin1(byte[] buf, int charPos, int c) {
+        UNSAFE.putByte(buf, ARRAY_BYTE_BASE_OFFSET + (long) charPos, (byte) c);
+    }
+
+    private static void putCharUTF16(byte[] buf, int charPos, int c) {
+        UNSAFE.putChar(buf, ARRAY_BYTE_BASE_OFFSET + ((long) charPos << 1), (char) c);
```

Comment:
I'm not sure we can put a `char` into a `byte[]`.
@cl4es is this safe on all platforms?

Reply:
The doc of `Unsafe::putChar()` delegates to the doc of `Unsafe::putInt()` which clearly states that the `Object` and `offset` arguments must locate a variable of the same type as the one of argument `x`, which is not the case here.

Reply:
This is safe. See https://github.com/openjdk/jdk/blob/3804082cba56e6d26c500880cc5cbe6d4332d8f8/src/java.base/share/classes/java/lang/invoke/X-VarHandleByteArrayView.java.template#L129-L132, where `MethodHandles.byteArrayViewVarHandle` implementation does getChar/Short/Int/Long on a byte array.

Reply:
The difference there is the use of `Unsafe.get/putCharUnaligned`. I think `putChar` might work for all platforms we're always storing at an index that is divisible by 2 in the byte array, thus aligned up to `char`. That said this might be an abuse of an unspecified behavior and adding `-Unaligned` might be a bit safer.